core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -134,54 +102,41 @@ public Object visitPrepareCommand(TxInvocationContext ctx, PrepareCommand comman
    }
 
    @Override
-   protected <K> Object handleWriteManyCommand(InvocationContext ctx, FlagAffectedCommand command, Collection<K> keys, boolean forwarded) throws Throwable {
-      Object maybeStage;
-      if (hasSkipLocking(command)) {
-         maybeStage = invokeNext(ctx, command);
+   protected <K> Object handleWriteManyCommand(InvocationContext ctx, FlagAffectedCommand command, Collection<K> keys,
+         boolean forwarded, int topologyId) throws Throwable {
+      final TxInvocationContext<?> txCtx = (TxInvocationContext<?>) ctx;
+      if (skipLockingForKeys(command, txCtx, keys)) {
+         txCtx.addAllAffectedKeys(keys);
+         return invokeNext(ctx, command);
+      } else if (isLocalOnly(command)) {
+         lockAllOrRegisterBackupLock(txCtx, keys, getLockTimeoutMillis(command));
+         txCtx.addAllAffectedKeys(keys);
+         return invokeNext(ctx, command);
       } else {
-         if (!needRemoteLocks(ctx, keys, command)) {
-            acquireLocalLocks(ctx, command, keys);
-            maybeStage = invokeNext(ctx, command);
-         } else {
-            final TxInvocationContext txContext = (TxInvocationContext) ctx;
-            LockControlCommand lcc = cf.buildLockControlCommand(keys, command.getFlagsBitSet(),
-                  txContext.getGlobalTransaction());
-            maybeStage = invokeNextThenApply(ctx, lcc, (rCtx, rCommand, rv) -> {
-               acquireLocalLocks(rCtx, command, keys);
-               return invokeNext(rCtx, command);
-            });
-         }
+         return invokeLockCommandAndContinue(keys, command, txCtx, topologyId);
       }
-      return maybeStage;
    }
 
    @Override
    protected Object visitDataWriteCommand(InvocationContext ctx, DataWriteCommand command)
          throws Throwable {
-      Object maybeStage;
-      Object key = command.getKey();
-      if (hasSkipLocking(command)) {
+      if (!ctx.isInTxScope()) {
+         //we don't know how to lock keys in non-tx contexts
```

Comment:
PFER needs to be locked the same way as in non-tx mode - lock it here and unlock after `invokeNext`.

Reply:
PFER is invoking `visitNonTxDataWriteCommand` and not this method.

Reply:
OK, is there any other way to get non-tx write command here? (If in doubt I usually throw exception and if testsuite doesn't fail, it's good to rule out one weird special case with error).

Reply:
you are the one who added the if-condition (https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java#L164) IMO, it doesn't make sense to have non-tx commands here...

Reply:
I don't recall why I did that; could you replace `invokeNext` with `throw new IllegalStateException()`?

Reply:
changing to `assert` if you don't mind